### DISCUSSION TITLE
How to replace the TextBox in a ComboBox with an instance of a derived class?

### DISCUSSION THREAD

**Discussion by trebahl** [dotnet/wpf#9999] [Q&A]
I have implemented an autocomplete behaviour for textboxes which works by overriding OnTextInput and OnKeyDown. I can't just subscribe to events because I need to perform stuff before the textbox's built in handlers.

Now I would like to plug this behaviour into the editable part of a combobox. That would require me to force the template of the combobox to create an instance of my derived textbox instead of the vanilla textbox. What would be the easiest way of doing that?

PS: I have devised an alternative way to overriding OnTextInput and OnKeyDown, but it's horrifying: I add to TextBox a class handler for the corresponding events and it gets executed before the handler installed by UIElement, but of course it will affect all textboxes and not just the ones I want to plug my behaviour into (I work around that by setting on the textboxes I want to autocomplete a special property value that I test in the class handler). I'd very much like to avoid using that method.

**Comment by trebahl** [comment]
I realize I can just use the preview events to execute stuff before the class's handler.

But I still would find an answer to my question interesting. More generally, being able to programmatically modify the template that is about to get applied would be great. At the moment, it feels to me like the promise of being able to customize a control's appearance using  templates is fulfilled in name only.

**Reply by miloush** [reply]
Not only you can use preview events, but also if you overriding should give you opportunity to run stuff before/instead of the built-in handlers, unless of course these are executed before.

Template is a DP, so you might be able to override its metadata and coerce the value. There is also [OnTemplateChanged](https://learn.microsoft.com/en-us/dotnet/api/system.windows.controls.control.ontemplatechanged) which is practically the property's changed handler. Either way you need to do the swap before ApplyTemplate is called (because that's were ComboBox gets a reference to it).

> it will affect all textboxes

You can workaround this by finding the PART_EditableTextBox on ApplyTemplate yourself and then check in the event handlers whether the text box to execute them for is the same.

**Reply by trebahl** [reply]
> Template is a DP, so you might be able to override its metadata and coerce the value

Yes, but templates from xaml are a black box so I can only replace them with one that I have to create from scratch. It would be nice to have an intermediate object model that could be edited by code.

**Reply by miloush** [reply]
You can probably remove the textbox from its parent and add yours instead, but I think the whole idea of replacing an element in a template is a bit flawed, it can hardly work in general case. The whole point of templates is that you have this separation. Could you not attach the autocompletion behavior to a plain TextBox instead of deriving from it?

Also, taking step back, what is the scenario here? Any autocomplete behavior for the textbox sounds like it would interfere with the ComboBox' autocomplete behavior...

**Reply by trebahl** [reply]
> the whole idea of replacing an element in a template is a bit flawed, it can hardly work in general case

In my case, I want to replace an element by an instance of a derived class, it's hard to imagine a non-contrived scenario where that would break something.

> Any autocomplete behavior for the textbox sounds like it would interfere with the ComboBox' autocomplete behavior...

Do you mean what you get with IsTextSearchEnabled = true? I have disabled it.

I have implemented something similar to what we had in winforms (pops up a listbox with the list of completions that match the text that has already been typed), with an option for file path completion. It's much better imho.

**Reply by miloush** [reply]
But you are lucky you are trying to replace a well-known part of the template, otherwise you wouldn't be able to tell which of the elements you want to replace.

> pops up a listbox with the list of completions

Kind of sounds like ComboBox to me, just without the dropdown arrow... can you not replace the whole control with yours? What about when the dropdown of ComboBox is open and you type in the textbox with its own dropdown? And who gets the keyboard events to open the popup?

Anyway, I don't think I have an answer you would be happy with, perhaps someone else has better ideas. Visual Studio let's you edit the ComboBox template, so you can replace the TextBox in there.

**Reply by trebahl** [reply]
> Kind of sounds like ComboBox to me, just without the dropdown arrow

Yes and no. We have a use case where we use such a combobox to let the user enter file paths. The items in the combobox are the previously entered files, and the autocomplete in the textbox is based on the filesystem instead of the items of the combobox.

This was possible in windows forms with standard options, I'm not inventing something crazy. I'm currently working to port my company code from windows forms to wpf, so I had to reimplement it.

>  What about when the dropdown of ComboBox is open and you type in the textbox with its own dropdown

When you type it hides the combobox dropdown, and when you open the combobox drop down it hides the autocomplete drop down if it was opened.
